trin-core/src/types/receipts.rs

```diff
@@ -0,0 +1,523 @@
+            Err(msg) => Err(anyhow!("Error calculating receipts root: {msg:?}")),
+        }
+    }
+}
+
+// Based on https://github.com/openethereum/openethereum/blob/main/crates/ethcore/types/src/receipt.rs
+
+/// A record of execution for a `LOG` operation.
+#[derive(Default, Debug, Clone, PartialEq, Eq, RlpEncodable, RlpDecodable)]
+pub struct LogEntry {
+    /// The address of the contract executing at the point of the `LOG` operation.
+    pub address: Address,
+    /// The topics associated with the `LOG` operation.
+    pub topics: Vec<H256>,
+    /// The data associated with the `LOG` operation.
+    pub data: Bytes,
+}
+
+impl LogEntry {
+    /// Calculates the bloom of this log entry.
+    pub fn bloom(&self) -> Bloom {
+        self.topics.iter().fold(
+            Bloom::from(BloomInput::Raw(self.address.as_bytes())),
+            |mut b, t| {
+                b.accrue(BloomInput::Raw(t.as_bytes()));
+                b
+            },
+        )
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TransactionOutcome {
+    /// Status and state root are unknown under EIP-98 rules.
+    Unknown,
+    /// State root is known. Pre EIP-98 and EIP-658 rules.
+    StateRoot(H256),
+    /// Status code is known. EIP-658 rules.
+    StatusCode(u8),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LegacyReceipt {
+    pub cumulative_gas_used: U256,
+    pub log_bloom: Bloom,
+    pub logs: Vec<LogEntry>,
+    pub outcome: TransactionOutcome,
+}
+
+impl LegacyReceipt {
+    pub fn decode_rlp(rlp: &Rlp) -> anyhow::Result<Self> {
+        match rlp.item_count()? {
+            3 => Ok(Self {
+                outcome: TransactionOutcome::Unknown,
+                cumulative_gas_used: rlp.val_at(0)?,
+                log_bloom: rlp.val_at(1)?,
+                logs: rlp.list_at(2)?,
+            }),
+            4 => Ok(Self {
+                cumulative_gas_used: rlp.val_at(1)?,
+                log_bloom: rlp.val_at(2)?,
+                logs: rlp.list_at(3)?,
+                outcome: {
+                    let first = rlp.at(0)?;
+                    if first.is_data() && first.data()?.len() <= 1 {
+                        TransactionOutcome::StatusCode(first.as_val()?)
+                    } else {
+                        TransactionOutcome::StateRoot(first.as_val()?)
+                    }
+                },
+            }),
+            _ => Err(anyhow!("Invalid encoded rlp length.")),
+        }
+    }
+}
+
+impl LegacyReceipt {
+    pub fn new(
+        outcome: TransactionOutcome,
+        cumulative_gas_used: U256,
+        logs: Vec<LogEntry>,
+    ) -> Self {
+        LegacyReceipt {
+            cumulative_gas_used,
+            log_bloom: logs.iter().fold(Bloom::default(), |mut b, l| {
+                b.accrue_bloom(&l.bloom());
+                b
+            }),
+            logs,
+            outcome,
+        }
+    }
+    pub fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
```

Comment:
Is there a reason that we aren't implementing the [`Decodable`](https://docs.rs/rlp/0.5.1/rlp/trait.Decodable.html) and [`Encodable`](https://docs.rs/rlp/0.5.1/rlp/trait.Encodable.html) traits here?